src/controls/controls.js

```diff
@@ -294,21 +318,57 @@ Crafty.c("Multiway", {
     },
 
     __applyActiveDirections: function() {
+        var vx = 0, vy = 0, directions = [];
         for (var direction in this._activeDirections) {
             if (this._activeDirections[direction] > 0) {
-                this.vx += this._directionSpeed[direction].x;
-                this.vy += this._directionSpeed[direction].y;
+                vx += this._directionSpeed[direction].x;
+                vy += this._directionSpeed[direction].y;
+                directions.push(direction);
             }
         }
+        if ((vx === 0) && (vy === 0)) return;
+        if (this._clampSpeed === true) {
+          var avgAngle = this.__directionsToRad(directions);
+          this.vx += Math.round(Math.abs(vx) * Math.cos(avgAngle) * 1000) / 1000;
+          this.vy += Math.round(Math.abs(vy) * Math.sin(avgAngle) * 1000) / 1000;
+        } else {
+          this.vx += vx;
+          this.vy += vy;
+        }
     },
 
     __unapplyActiveDirections: function() {
+        var vx = 0, vy = 0, directions = [];
         for (var direction in this._activeDirections) {
             if (this._activeDirections[direction] > 0) {
-                this.vx -= this._directionSpeed[direction].x;
-                this.vy -= this._directionSpeed[direction].y;
+                vx += this._directionSpeed[direction].x;
+                vy += this._directionSpeed[direction].y;
+                directions.push(direction);
             }
         }
+        if ((vx === 0) && (vy === 0)) return;
+        if (this._clampSpeed === true) {
+          var avgAngle = this.__directionsToRad(directions);
+          this.vx -= Math.round(Math.abs(vx) * Math.cos(avgAngle) * 1000) / 1000;
+          this.vy -= Math.round(Math.abs(vy) * Math.sin(avgAngle) * 1000) / 1000;
+        } else {
+          this.vx -= vx;
+          this.vy -= vy;
+        }
+    },
+
+    __directionsToRad: function(directions) {
```

Comment:
IIRC I  looked at this approach before and ran into some problems.  The average of even two angles it ill-defined:  if the inputs are left (280) and right (0), you could get either 90 or 270 degrees as the average.  And neither is what you'd intuitively expect from the averaged input, which would simply be the zero-vector.

Instead of averaging the angle, I think you want to take the existing implementation and normalize the resultant vector.  You'd have an additional pass that would be something like:
``` javascript
var unnormed = sqrt(this.vx * this.vx + this.vy * this.vy);
this.vx = this.vx * (this._clampSpeed / unnormed);
this.vy = this.vy * (this._clampSpeed / unnormed);
```